### PR TITLE
Refactor BackendServiceException Sentry logging

### PR DESCRIPTION
**Pull request by kfrz**
## Description of change
As a developer I want backend service exceptions to use structured logging in CloudWatch so that I can more easily create queries.

#### Before:
screenshot-user-images githubusercontent com-2019 01 23-05-06-16

#### After:
screenshot-user-images githubusercontent com-2019 01 23-05-10-21

This payload should filter through to Cloudwatch in a more useful manner, allowing for searching by status or source. 

## Testing done
Checked the log output by `tail -f log/test.log` while running test suite. 

Used strategy employed by @annaswims [here](https://github.com/department-of-veterans-affairs/vets-api/pull/2723) to override logging and ensure correct output. 

#### Unique to this PR
- [x] Log additional data as separate keys in the payload rather than interpolated them in the message string. 

#### Applies to all PRs
- [ ] Appropriate logging
- [ ] Swagger docs have been updated, if applicable
- [ ] Provide link to originating GitHub issue, or connected to it via ZenHub
- [ ] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [ ] Provide which alerts would indicate a problem with this functionality (if applicable)

### CODE REVIEW

**Comment by kfrz** [comment]
@saneshark We **can** coerce the message sent to Sentry by adding a processor, and filtering by exception type as we are doing in [`lib/sentry/processor/log_as_warning.rb`](https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/sentry/processor/log_as_warning.rb#L19) but the issue here is I think we don't have access to the exception class - just the message itself. We could parse the stringified message from the `data` blob but that doesn't seem like a good solution to me, it's brittle and feels forced.

I think we should open a parallel ticket and track the Sentry processor work there - this PR should only affect the Rails logs as per #15653.

What I would like to find a way to accomplish, and haven't been able to, is to inject `exception.errors` into the `data` hash from Raven. 

Any other ideas in this regard?

**Comment by kfrz** [comment]
@kreek Is there a way to test/verify the Cloudwatch logging structure on a review instance or maybe dev/staging? I would think we would have to trigger some sort of exception to be raised - should doing this at the console work to trigger messages into Cloudwatch?

**Comment by kreek** [comment]
@kfrz Anna had a PR where she wanted to check production style logs locally, her instructions are [here](https://github.com/department-of-veterans-affairs/vets-api/pull/2723). Not sure if we keep review instance logs in AWS, we do have dev and staging logs so you could merge after the automated deploy, test, and roll-back if necessary. I'll check on your AWS permissions.

File: lib/sentry_logging.rb [hunk 1]
```diff
@@ -24,9 +26,16 @@ def log_exception_to_sentry(
...
       Raven.capture_exception(exception.cause.presence || exception, level: level)
     end
-    rails_logger(level, "#{exception.message}.")
+
+    if exception.is_a? Common::Exceptions::BackendServiceException
+      rails_logger(level, exception.message, exception.errors)
+    else
+      rails_logger(level, "#{exception.message}.")
+    end
     rails_logger(level, exception.backtrace.join("\n")) unless exception.backtrace.nil?
```
[the remark this answers is not in the document]
Comment (kfrz) [reply lib/sentry_logging.rb]: @kreek

```rb
rails_logger(level, exception.backtrace.join("\n")) unless exception.backtrace.nil?
```

The above code would print the backtrace as one atomic log line - are you suggesting something like:

```rb
    if errors.present?
     payload = errors.merge(backtrace: backtrace)
     rails_logger(level, exception.message, payload)
    else
      rails_logger(level, "#{exception.message}.", {backtrace: backtrace})
    end
```

Effectively making a request with an exception print as one big fat logger line - is this what you'd be aiming for with this refinement?

I wonder, if log lines could be grouped by request id would this still be an impactful change?

[the remark this answers is not in the document]
Comment (kreek) [reply lib/sentry_logging.rb]: Log lines can be grouped by request id. However I think as we have structured logging in place all the data relating to an event should be on one line. I wouldn't worry about the size of a line, tools like awslogs tool let you [filter fields](https://github.com/jorgebastida/awslogs#json-logs) so you can see only the message field if you want to.

[the remark this answers is not in the document]
Comment (kfrz) [reply lib/sentry_logging.rb]: @kreek What's stopping us from doing simply replacing this whole block:

```rb
    if exception.is_a? Common::Exceptions::BackendServiceException
      rails_logger(level, exception.message, exception.errors, exception.backtrace)
    else
      rails_logger(level, "#{exception.message}.")
    end
    rails_logger(level, exception.backtrace.join("\n")) unless exception.backtrace.nil?
```

with this single line:

```rb
  rails_logger(level, exception.message, exception.errors, exception.backtrace)
```

The latter would include all information on a single line including backtrace or errors. Maybe there's a Good Reason that I'm not considering  - thoughts?

**Pull request merged**